File: solutions/javascript-solutions/src/problems/1-two-sum.js
import { binarySearch } from "../utils.js";

/**
 * @param {number[]} nums 
 * @param {number} target 
 * @returns {number[]}
 */
export function twoSumBrute(nums, target) {
    for (let i = 0; i < nums.length - 1; i++)
        for (let j = i + 1; j < nums.length; j++)
            if (nums[i] + nums[j] == target)
                return [i, j];
    throw new Error('No solution found');
}

/**
 * @param {number[]} nums 
 * @param {number} target 
 * @returns {number[]}
 */
export function twoSumSort(nums, target) {
    let numsWithIndices = [];
    for (let i = 0; i < nums.length; i++)
        numsWithIndices.push({ value: nums[i], index: i });
    numsWithIndices.sort((a, b) => a.value - b.value);
    for (let i = 0; i < nums.length; i++) {
        const desired = target - numsWithIndices[i].value;
        const found = binarySearch(numsWithIndices, { value: desired }, i + 1, nums.length - 1, (a, b) => a.value - b.value);
        if (found > 0)
            return [numsWithIndices[i].index, numsWithIndices[found].index];
    }
    throw new Error('No solution found');
}

/**
 * @param {number[]} nums 
 * @param {number} target 
 * @returns {number[]}
 */
export function twoSumHash(nums, target) {
    const processedNums = new Map();
    for (let i = 0; i < nums.length; i++) {
        const j = processedNums.get(target - nums[i]);
        if (j !== undefined)
            return [j, i];
        processedNums.set(nums[i], i);
    }
    throw new Error('No solution found');
}

File: solutions/java-solutions/src/main/java/com/mikkkkkkka/problems/_1TwoSum.java
package com.mikkkkkkka.problems;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class _1TwoSum {
    public static int[] twoSumBrute(int[] nums, int target) {
        for (int i = 0; i < nums.length - 1; i++)
            for (int j = i + 1; j < nums.length; j++)
                if (nums[i] + nums[j] == target)
                    return new int[] { i, j };
        throw new IllegalArgumentException("No solution found");
    }

    private record NumAndIndex(int value, int index) {
        public static int Comparator(NumAndIndex a, NumAndIndex b) {
            return a.value - b.value;
        }
    }

    public static int[] twoSumSort(int[] nums, int target) {
        var numsWithIndex = new NumAndIndex[nums.length];
        for (int i = 0; i < nums.length; i++)
            numsWithIndex[i] = new NumAndIndex(nums[i], i);
        Arrays.sort(numsWithIndex, NumAndIndex::Comparator);
        for (int i = 0; i < nums.length; i++) {
            int desired = target - numsWithIndex[i].value;
            int found = Arrays.binarySearch(numsWithIndex, i + 1, nums.length, new NumAndIndex(desired, 0),
                    NumAndIndex::Comparator);
            if (found > 0)
                return new int[] { numsWithIndex[i].index, numsWithIndex[found].index };
        }
        throw new IllegalArgumentException("No solution found");
    }

    public static int[] twoSumHash(int[] nums, int target) {
        Map<Integer, Integer> processedNums = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int desired = target - nums[i];
            if (processedNums.containsKey(desired))
                return new int[] { processedNums.get(desired), i };
            processedNums.put(nums[i], i);
        }
        throw new IllegalArgumentException("No solution found");
    }
}
